### PR TITLE
[CoSim] access to SubSolver

### PR DESCRIPTION
**Pull request by philbucher**
Adding functionalities for accessing the SubSolvers in a CoSimulation
This is needed in order to access the `Model`s of the corresponding solvers
for now WIP, @KlausBSautter please test if this works for you, I haven't tested it.
Test will be added later
BTW this is a hotfix in its current state, suggestions for improvement and better names are appreciated

In the near future and as preparation for MultiStageAnalysis we also need to enhance the `CoSimulationAnalysis` such that it can get a `Model` from outside (This will be probably a key-value pair passed in the constructor of the `CoSimulationAnalysis`, where the Key is the (Full) name  of the Sovler and the Value is the `Model` the solver should use)

### CODE REVIEW

File: applications/CoSimulationApplication/python_scripts/co_simulation_analysis.py [hunk 1]
```diff
@@ -90,6 +90,9 @@ def OutputSolutionStep(self):
         if self.flush_stdout:
             CoSimulationAnalysis.Flush()
 
+    def GetSubSolver(self, sub_solver_name):
+        self._GetSolver(solver_name).GetSubSolver(sub_solver_name)
```
Comment (KlausBSautter) [inline applications/CoSimulationApplication/python_scripts/co_simulation_analysis.py]:
```suggestion
        return self._GetSolver().GetSubSolver(sub_solver_name)
```

**Comment by KlausBSautter** [comment]
---
        dem = self.GetSubSolver("dem")
        fem = self.GetSubSolver("structure")

finds the solver but returns a `NoneType`

**Comment by philbucher** [comment]
Ok then lets take a look together tomorrow

**Comment by KlausBSautter** [comment]
thanks, works for me

**Comment by adityaghantasala** [comment]
A couple of questions : 
- What is a sub-solver ?
- Where is the intended use of this function `GetSubSolver` ?

Secondly, I understand why do you want to pass a `Model` from outside to the co-sim analysis. In that case will the `Solvers` use this `Model` ? or make their own as what is currently done?

**Comment by KlausBSautter** [comment]
@adityaghantasala I use this function to check and save nodal results from the respective solvers
(in a DEM-FEM coupled simulation DEM&FEM are the subsolvers

-----
class my_analysis(CoSimulationAnalysis):

    def OutputSolutionStep(self):
        super(my_analysis,self).OutputSolutionStep()

        dem = self.GetSubSolver("dem").model
        fem = self.GetSubSolver("structure").model

        time_vec.append(self.GetSubSolver("structure")._analysis_stage.time)
        temp_reac_z = 0
        for node_i in fem["Structure.DISPLACEMENT_dirichlet"].Nodes:
            temp_reac_z+=node_i.GetSolutionStepValue(KratosMultiphysics.REACTION_Z)
        struct_reaction_vec.append(temp_reac_z)



        for node_i in dem["SpheresPart.Parts_particle"].Nodes:
            particle_disp.append(node_i.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Z))
            particle_vel.append(node_i.GetSolutionStepValue(KratosMultiphysics.VELOCITY_Z))

**Comment by philbucher** [comment]
@adityaghantasala those things were done in a rush, hence the draft :) Now let me answers your questions:

>     * What is a sub-solver ?

Some solver inside of CoSim. I.e. in a regular GS-FSI this would be `FluidDynamicsWrapper` or `StructuralMechanicsWrapper` (or `DEMWrapper` in the case of @KlausBSautter)
=> we need some mechanism to access those, e.g. for what @KlausBSautter wants to do
BTW I am not really happy with the naming, if you have a better suggestion I'd be happy :)

>     * Where is the intended use of this function `GetSubSolver` ?

See e.g. what @KlausBSautter is doing

> Secondly, I understand why do you want to pass a `Model` from outside to the co-sim analysis. In that case will the `Solvers` use this `Model` ? or make their own as what is currently done?

In case no `Model` is passed from outside, they will use their own, as is currently done. In case it is being passed from outside, then they will use this

**Comment by philbucher** [comment]
actually what @KlausBSautter wants to do can also be done if we can pass in `Model`s from outside, but I still think it can be beneficial to also have access to the "SubSolver"s

**Comment by adityaghantasala** [comment]
@KlausBSautter  
In the above 
`struct_reaction_vec` , `particle_disp` and `particle_vel` are used in some calculations later I guess ? 

Saying that @philbucher  Can't such a calculation be done via `CouplingOperation`s ?

> In case no `Model` is passed from outside, they will use their own, as is currently done. In case it is being passed from outside, then they will use this

This can cause problems as we cannot guarantee unique names for modelparts from each solver. Imagine fluid-fluid or str-str interaction, both the modelparts can be name FluidModelPart or MainModelPart (which is the default one from GiD output). 


> BTW I am not really happy with the naming, if you have a better suggestion I'd be happy :)

well `SubSolver`  does not sound consistent, also this terminology is not used so far. So using the existing names/terminology just call it `GetSolver` .

**Comment by philbucher** [comment]
> @KlausBSautter
> In the above
> `struct_reaction_vec` , `particle_disp` and `particle_vel` are used in some calculations later I guess ?
> 
> Saying that @philbucher Can't such a calculation be done via `CouplingOperation`s ?

Yes it can be done like that. Not sure though what makes more sense. @KlausBSautter wants to do some postprocessing and this is maybe easier in the Main
@KlausBSautter what do you think

> > In case no `Model` is passed from outside, they will use their own, as is currently done. In case it is being passed from outside, then they will use this
> 
> This can cause problems as we cannot guarantee unique names for modelparts from each solver. Imagine fluid-fluid or str-str interaction, both the modelparts can be name FluidModelPart or MainModelPart (which is the default one from GiD output).

I think you misunderstood, ofc for the reasons you mention we still have one `Model` per `SolverWrapper`.

> > BTW I am not really happy with the naming, if you have a better suggestion I'd be happy :)
> 
> well `SubSolver` does not sound consistent, also this terminology is not used so far. So using the existing names/terminology just call it `GetSolver` .

So you suggest to override `_GetSolver`? I like the idea!

**Comment by philbucher** [comment]
@adityaghantasala this is how I imagine the handling of the `Model`s in MultiStageAnalysis
Imagine an FSI-Analysis with membranes that need formfinding prior to the FSI:
~~~py
# Stage 1: FormFinding:
model_str = KM.Model()
StructuralMechanicsAnalysis(model_str, proj_params_formfinding).Run()

co_sim_model_dict = {
    "structure" : model_str # "structure" is the name in the "proj_params_fsi" 
                            # for the Structural Solver
}

# Stage 2: FSI
# IMPORTANT:
# the fluid solver in CoSim does NOT get a Model from outside
# hence it creates its own!
CoSimulationAnalysis(proj_params_fsi, co_sim_model_dict).Run() # run FSI



~~~

**Comment by adityaghantasala** [comment]
If it can be done with coupling operations (which I believe it can be), I will strongly encourage using this way.

@philbucher  about multistage analysis, that look nice, but may be we can '`some how`' incorporate it into co-sim workflow. As in it looks like another IO operation :-| 

Nevertheless  we will require` _GetSolver` at some point, but I am really cautious because it can lead to abuse of that function.

**Comment by philbucher** [comment]
> @philbucher about multistage analysis, that look nice, but may be we can 'some how' incorporate it into co-sim workflow. As in it looks like another IO operation :-|

Can you explain a bit, I don't fully understand

**Comment by KlausBSautter** [comment]
what is the current state of this?

**Comment by philbucher** [comment]
I need to finish the test, otherwise is ready

**Comment by philbucher** [comment]
I added the tests

I also took the change to update the Import mechanism and add some more checks in the `CoupledSolver`

**Review by KlausBSautter — DISMISSED** [review]
👍

**Review by KlausBSautter — APPROVED** [review]

**Pull request merged**